### DISCUSSION TITLE
Help on ContactSplit

### DISCUSSION THREAD

**Discussion by matthiasneuner** [idaholab/moose#17501] [Q&A Modules: General] [answered]
Hi,

I am currently trying to set up a simulation using the `ContactSplit` module for preconditioning a contact module.
However, I end up with following PETSc message:
```
[0]PETSC ERROR: --------------------- Error Message --------------------------------------------------------------
[0]PETSC ERROR: Petsc has generated inconsistent data
[0]PETSC ERROR: Unhandled case, must have at least two fields, not 0
```

My `Contact` and `ContactSplit` definition look like:
```

[Preconditioning]
  active='contact_fsp'

  [contact_fsp]
      type = FSP
      topsplit = 'contact_interior' 
      [contact_interior]
        splitting = 'contact interior'
        splitting_type = multiplicative
      []
      [./interior]
          type = ContactSplit
          vars = 'disp_x disp_y disp_z'
          uncontact_primary   = 'set_hammer_1'
          uncontact_secondary    = 'set_channel_1'
          uncontact_displaced = '1'
          include_all_contact_nodes = 1
          petsc_options_iname = '-ksp_type -ksp_max_it -ksp_rtol -ksp_gmres_restart -pc_type -pc_hypre_type -pc_hypre_boomeramg_max_iter -pc_hypre_strong_threshold'
          petsc_options_value = ' preonly 10 1e-4 201                hypre    boomeramg      1                            0.25'
      [../]
      [./contact]
          type = ContactSplit
          vars = 'disp_x disp_y disp_z'
          contact_primary   = 'set_hammer_1'
          contact_secondary    = 'set_channel_1'
          contact_displaced = '1'
          include_all_contact_nodes = 1
          petsc_options_iname = '-ksp_type -ksp_max_it -pc_type -pc_asm_overlap -sub_pc_type   -pc_factor_levels'
          petsc_options_value = '  preonly 10 asm  1 lu 0'
      [../]
  []

  [smp_strumpack]
    type = SMP
    full = true
    petsc_options_iname = '-pc_type -pc_solver_type'
    petsc_options_value = 'lu strumpack'
  []

[]

[Contact]
  [1]
    formulation = penalty
    primary = set_hammer_1
    secondary = set_channel_1
    model = frictionless
    penalty = 1e+6
  []
[]
```

which I actually implemented following the testcases from the contact module.
I am not sure what the options `contact_displaced = '1'` and `uncontact_displaced = '1'` should do, but deactivating them doesn't resolve any issue.

Please, can somebody give a hint on the correct implementation of `ContactSplit`?

€: Please find the files to reproduce the error here: https://fileshare.uibk.ac.at/d/ce29dee42ca04aa3a7e3/

**Comment by jiangwen84** [comment]
@dewenyushu I remembered you used `ContactSplit` before, and could you help answer the question?

**Comment by dewenyushu** [comment]
Hi @matthiasneuner 

I need to admit that this problem seems to be a more complicated case than what I've played with before. Let's see if I may be of some help to you:

By looking at your mesh, I realized that you actually have two sets of contact surfaces, i.e., set_hammer_1 in contact with set_channel_1, and set_hammer_2 in contact with set_channel_2. However, the sideset 'set_hammer_2' is missing in your mesh thus your FSP block. 

Since FSP uses the sideset info to extract the degree of freedoms and apply different preconditioners on the corresponding sub-problems, I would suggest making this correction in your mesh and add the second pair of contact surfaces in FSP. 

I am not sure about the  `contact_displaced ` and `uncontact_displace ` usage here though. @fdkong do you have some suggestions?

**Reply by matthiasneuner** [reply]
Hello, thank you for the response.

Indeed, I want to include later on a second contact pair (set_hammer_2/set_channel_2). However, currently it has no contact definition assigned, so only (set_hammer_1/set_channel_1) is active at the moment. 
But this should not be the cause of the problem?

**Reply by dewenyushu** [reply]
Oh, yeah this should not be the cause of the problem then. Would you mind creating a new issue for this? I suspect this example contains some corner cases that are not properly taken care of in the FSP implementation.

**Comment by fdkong** [comment]
> I am not sure about the contact_displaced and uncontact_displace usage here though. @fdkong do you have some suggestions?

`contact_displaced` and `uncontact_displace` set a flag to check if we want to use a displaced mesh or not. If use the displaced mesh, geometric search information from the displaced mesh will be taken. 

```
            if (displaced)
            {
              std::shared_ptr<DisplacedProblem> displaced_problem =
                  dmm->_nl->_fe_problem.getDisplacedProblem();
              if (!displaced_problem)
              {
                std::ostringstream err;
                err << "Cannot use a displaced contact (" << it.second.first << ","
                    << it.second.second << ") with an undisplaced problem";
                mooseError(err.str());
              }
              locator = displaced_problem->geomSearchData()._penetration_locators[it.first];
            }
            else
              locator = dmm->_nl->_fe_problem.geomSearchData()._penetration_locators[it.first];
```

**Comment by fdkong** [comment]
> [0]PETSC ERROR: --------------------- Error Message --------------------------------------------------------------
[0]PETSC ERROR: Petsc has generated inconsistent data
[0]PETSC ERROR: Unhandled case, must have at least two fields, not 0

It  seems all petsc options were  removed before calling PETSc, and PETSc got nothing from MOOSE, and could not set FSP

**Comment by fdkong** [comment]
You could take out of these blocks, and try again. Especially I did not use FSP with a `Predictor` block.

```
  [TimeStepper]
    type = IterationAdaptiveDT
    optimal_iterations = 8
    iteration_window = 3
    linear_iteration_ratio = 1000
    growth_factor = 1.5
    cutback_factor = 0.5
    dt = 1.0e-4
  []
  [Predictor]
    type = SimplePredictor
    scale = 1.0
    skip_after_failed_timestep = true
  []
```

**Reply by matthiasneuner** [reply]
Unfortunately, this does not resolve the issue.

**Comment by GiudGiud** [comment, marked as answer]
Has the issue been created? If so please link it to this discussion

**Reply by matthiasneuner** [reply]
Not yet; I had no time to investigate this further, and I am still not sure if it is not my fault. I will create an issue with a MWE in the next few days.

**Reply by fdkong** [reply]
Thanks. Right, it might not be your fault at all. However, a MWE will help us to get things fixed.

Thanks so much.

**Reply by GiudGiud** [reply]
Any news on this?

**Reply by fdkong** [reply]
@matthiasneuner Do you have  a MWE yet?

**Reply by matthiasneuner** [reply]
Hello, sorry for the long delay. I am trying to create a MWE, but somehow i fail to create a simple MWE which works even for a monolithic preconditioning. Please, can you tell me what is wrong with this simple MWE:
https://fileshare.uibk.ac.at/d/8f9f69e228674162b027/

? The contact never becomes active, and penetration occurs.

In any case, switching to  contact split basically invokes the same error initially discussed in this thread .. 

Thank you in advance!

**Reply by fdkong** [reply]
Create an issue here #18072

Will look into it when there is a chance
